### PR TITLE
Implement a MetagraphSyncer to sync multiple metagraphs

### PR DESCRIPTION
**Pull request by RusticLuftig**
At the moment, this change is a no-op since it's not integrated into the miner or vali. I'll integrate it in a separate change since it'll require a sizable refactoring of the vali.

### CODE REVIEW

File: common/metagraph_syncer.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        """Performs an initial sync of all metagraphs.
+
+        Unlike regular syncs, this will not notify listeners of the updated metagraph.
+        """
+        bt.logging.debug("Metagraph syncer do_initial_sync called")
+
+        for netuid in self.config.keys():
+            fn = functools.partial(self.subtensor.metagraph, netuid)
+            with self.lock:
+                metagraph = utils.run_in_subprocess(fn, ttl=120)
```
Comment (Sid-Data-Universe) [inline common/metagraph_syncer.py:49]: do we need the lock around the run_in_subprocess or can we have it just around the state update? As you have it below. I suppose we are blocking on one metagraph update either way so just a minor consistency nit.

File: common/metagraph_syncer.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            # On start, wait cadence before the first sync.
+            await asyncio.sleep(cadence)
+
+            try:
+                # Intentionally block the shared thread so that we only
+                # sync 1 metagraph at a time.
+                metagraph = utils.run_in_subprocess(
+                    functools.partial(self.subtensor.metagraph, netuid), ttl=120
+                )
+                state = None
```
Comment (Sid-Data-Universe) [inline common/metagraph_syncer.py:78]: Don't we immediately get state from the map in the next line? If that throws an exception then setting state to None here wouldn't help right?

File: common/metagraph_syncer.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                await asyncio.sleep(cadence)
+            except (BaseException, Exception) as e:
+                bt.logging.error(
+                    f"Error when syncing metagraph for {netuid}: {e}. Retrying in 60 seconds."
+                )
+                await asyncio.sleep(60)
+
+    def _run(self):
+        # For each netuid we should sync metagraphs for, spawn a Task to sync it.
+        asyncio.gather(
```
Comment (Sid-Data-Universe) [inline common/metagraph_syncer.py:96]: Should this just be wait if we aren't using the gathered return?

File: common/metagraph_syncer.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        """
+        self.subtensor = subtensor
+        self.config = config
+        self.metagraph_map: Dict[int, MetagraphSyncer._State] = {
+            netuid: MetagraphSyncer._State() for netuid in config.keys()
+        }
+        self.is_running = False
+        self.done_initial_sync = False
+        self.lock = threading.RLock()
+        self.notify_listener_thrad_pool = ThreadPoolExecutor(max_workers=1)
```
Comment (Sid-Data-Universe) [inline common/metagraph_syncer.py:37]: thrad->thread

**Review by Sid-Data-Universe — APPROVED** [review]

File: common/metagraph_syncer.py (hunk 1 above)
Comment (RusticLuftig) [reply common/metagraph_syncer.py:49]: Done

File: common/metagraph_syncer.py (hunk 2 above)
Comment (RusticLuftig) [reply common/metagraph_syncer.py:78]: State is accessed outside the lock on line 85, so I need to define it outside the lock to keep it in scope.

File: common/metagraph_syncer.py (hunk 3 above)
Comment (RusticLuftig) [reply common/metagraph_syncer.py:96]: Good catch - didn't realize there was a wait() API.

File: common/metagraph_syncer.py (hunk 4 above)
Comment (RusticLuftig) [reply common/metagraph_syncer.py:37]: But thrads are so much cooler...

**Review by Sid-Data-Universe — APPROVED** [review]

**Pull request merged**